In [ ]:
import torch
print(f"Python version: {torch.__version__}")
print(f"CUDA available: {torch.cuda.is_available()}")

Python version: 2.11.0+cu128
CUDA available: True


Get MACE working!

In [ ]:
# Clone the official MACE repository
!git clone https://github.com/ACEsuit/mace.git

# Install MACE from source in editable mode with Weights & Biases support
!pip install ./mace[wandb]

!pip install cuequivariance cuequivariance-torch cuequivariance-ops-torch-cu12

Cloning into 'mace'...
remote: Enumerating objects: 14696, done.
remote: Counting objects: 100% (258/258), done.
remote: Compressing objects: 100% (165/165), done.
remote: Total 14696 (delta 181), reused 93 (delta 93), pack-reused 14438 (from 3)
Receiving objects: 100% (14696/14696), 172.08 MiB | 34.06 MiB/s, done.
Resolving deltas: 100% (10961/10961), done.
Processing ./mace
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 387.7/387.7 kB 14.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 75.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 344.7/344.7 kB 33.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 453.1/453.1 kB 43.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 983.4/983.4 kB 70.4 MB/s eta 0:00:00
  Created wheel for mace-torch: filename=mac

In [ ]:
import wandb

wandb.login()

/usr/local/lib/python3.12/dist-packages/notebook/notebookapp.py:191: SyntaxWarning: invalid escape sequence '\/'
  | |_| | '_ \/ _` / _` |  _/ -_)
wandb: (1) Create a W&B account
wandb: (2) Use an existing W&B account
wandb: (3) Don't visualize my results


wandb: Enter your choice: 2


wandb: You chose 'Use an existing W&B account'
wandb: Logging into https://api.wandb.ai. (Learn how to deploy a W&B server locally: https://wandb.me/wandb-server)
wandb: Create a new API key at: https://wandb.ai/authorize?ref=models
wandb: Store your API key securely and do not share it.


wandb: Paste your API key and hit enter: ··········


wandb: No netrc file found, creating one.
wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc
wandb: Currently logged in as: m-abirhossain619 (m-abirhossain619-department-of-physics-dhaka-university-) to https://api.wandb.ai. Use `wandb login --relogin` to force relogin


True

In [ ]:
!wget -O train.xyz https://raw.githubusercontent.com/xke123502/PE_MACE/764ea602/example/train.xyz
!wget -O test.xyz  https://raw.githubusercontent.com/xke123502/PE_MACE/764ea602/example/test.xyz

--2026-08-04 20:45:29--  https://raw.githubusercontent.com/xke123502/PE_MACE/764ea602/example/train.xyz
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.110.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1594902 (1.5M) [text/plain]
Saving to: ‘train.xyz’

train.xyz           100%[===================>]   1.52M  --.-KB/s    in 0.04s   

2026-08-04 20:45:29 (34.5 MB/s) - ‘train.xyz’ saved [1594902/1594902]

--2026-08-04 20:45:29--  https://raw.githubusercontent.com/xke123502/PE_MACE/764ea602/example/test.xyz
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.108.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 418635 (409K) [text/plain]
Sa

Training MACE model on a mock data in `.xyz` format.

In [ ]:
!mace_run_train \
    --name="MACE_Research_Intro" \
    --train_file="train.xyz" \
    --valid_fraction=0.1 \
    --energy_key="energy" \
    --forces_key="forces" \
    --E0s="average" \
    --model="MACE" \
    --hidden_irreps="128x0e + 128x1o" \
    --r_max=5.0 \
    --batch_size=4 \
    --max_num_epochs=5 \
    --device=cuda

/usr/local/lib/python3.12/dist-packages/e3nn/o3/_wigner.py:10: UserWarning: Environment variable TORCH_FORCE_NO_WEIGHTS_ONLY_LOAD detected, since the`weights_only` argument was not explicitly passed to `torch.load`, forcing weights_only=False.
  _Jd, _W3j_flat, _W3j_indices = torch.load(os.path.join(os.path.dirname(__file__), 'constants.pt'))
/usr/local/lib/python3.12/dist-packages/torch/jit/_script.py:1488: DeprecationWarning: `torch.jit.script` is deprecated. Please switch to `torch.compile` or `torch.export`.
  warnings.warn(
2026-08-02 17:47:12.920 INFO: ===========VERIFYING SETTINGS===========
2026-08-02 17:47:12.920 INFO: MACE version: 0.3.17
2026-08-02 17:47:12.966 INFO: CUDA version: 12.8, CUDA device: 0
2026-08-02 17:47:13.004 INFO: ===========LOADING INPUT DATA===========
2026-08-02 17:47:13.004 INFO: Using heads: ['Default']
2026-08-02 17:47:13.004 INFO: Using the key specifications to parse data:
2026-08-02 17:47:13.004 INFO: Default: KeySpecification(info_keys={'energy': '

Test the model on mock test data in `.xyz` format.

In [ ]:
!mace_eval_configs \
    --configs="test.xyz" \
    --model="checkpoints/MACE_Research_Intro_run-123.model" \
    --output="test_eval.xyz" \
    --device=cuda \
    --batch_size=4

/usr/local/lib/python3.12/dist-packages/e3nn/o3/_wigner.py:10: UserWarning: Environment variable TORCH_FORCE_NO_WEIGHTS_ONLY_LOAD detected, since the`weights_only` argument was not explicitly passed to `torch.load`, forcing weights_only=False.
  _Jd, _W3j_flat, _W3j_indices = torch.load(os.path.join(os.path.dirname(__file__), 'constants.pt'))
/usr/local/lib/python3.12/dist-packages/torch/jit/_script.py:1488: DeprecationWarning: `torch.jit.script` is deprecated. Please switch to `torch.compile` or `torch.export`.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/mace/cli/eval_configs.py:160: UserWarning: Environment variable TORCH_FORCE_NO_WEIGHTS_ONLY_LOAD detected, since the`weights_only` argument was not explicitly passed to `torch.load`, forcing weights_only=False.
  model = torch.load(f=args.model, map_location=args.device)
/usr/local/lib/python3.12/dist-packages/torch/jit/_serialization.py:176: DeprecationWarning: `torch.jit.load` is deprecated. Please switch to `torch.expo

In [ ]:
!ls -la test_eval.xyz
!head -5 test_eval.xyz

-rw-r--r-- 1 root root 413634 Aug  2 18:04 test_eval.xyz
160
Lattice="9.7195997238 0.0 0.0 0.0 11.2231998444 0.0 0.0 0.0 50.0" Properties=species:S:1:pos:R:3:field_scalar:R:1:MACE_forces:R:3 MACE_energy=-934.239510876524 pbc="T T T"
Pt       0.00000000       0.00000000       1.75000000      -4.39613300      -0.01111957      -0.03596378      -0.16279158
Pt       2.42989993       1.40289998       1.75000000      -4.39613300       0.02397981      -0.01591690      -0.22886521
Pt       0.32277306       0.64354822       8.58661510      -4.39613300      -0.12792133      -0.15936739      -0.02028764


In [ ]:
!python -m mace.cli.preprocess_data --help

/usr/local/lib/python3.12/dist-packages/e3nn/o3/_wigner.py:10: UserWarning: Environment variable TORCH_FORCE_NO_WEIGHTS_ONLY_LOAD detected, since the`weights_only` argument was not explicitly passed to `torch.load`, forcing weights_only=False.
  _Jd, _W3j_flat, _W3j_indices = torch.load(os.path.join(os.path.dirname(__file__), 'constants.pt'))
/usr/local/lib/python3.12/dist-packages/torch/jit/_script.py:1488: DeprecationWarning: `torch.jit.script` is deprecated. Please switch to `torch.compile` or `torch.export`.
  warnings.warn(
usage: preprocess_data.py [-h] [--config CONFIG] --train_file TRAIN_FILE
                          [--valid_file VALID_FILE]
                          [--num_process NUM_PROCESS]
                          [--valid_fraction VALID_FRACTION]
                          [--test_file TEST_FILE] [--work_dir WORK_DIR]
                          [--h5_prefix H5_PREFIX] [--r_max R_MAX]
                          [--config_type_weights CONFIG_TYPE_WEIGHTS]
                  

In [ ]:
!ls -a

.  ..  .config	mace  sample_data  test.xyz  train.xyz


In [18]:
from ase.io import read, write

train_atoms = read("train.xyz", index=":")
test_atoms = read("test.xyz", index=":")

removed_count = 0
for atoms in train_atoms + test_atoms:
    if "field_scalar" in atoms.arrays:
        atoms.arrays.pop("field_scalar")
        removed_count += 1

print(f"Stripped field_scalar from {removed_count} configs")

write("train_clean.xyz", train_atoms, format="extxyz")
write("test_clean.xyz", test_atoms, format="extxyz")

Stripped field_scalar from 101 configs


In [26]:
from ase.io import read, write

train_atoms = read("train.xyz", index=":")
test_atoms = read("test.xyz", index=":")

for atoms in train_atoms + test_atoms:
    if "field_scalar" in atoms.arrays:
        del atoms.arrays["field_scalar"]

    # forces live on the calculator, not atoms.arrays -- pull them out first
    forces = atoms.get_forces()
    energy = atoms.get_potential_energy()

    # detach the calculator so nothing re-derives 'forces' key on write
    atoms.calc = None

    # add force as a proper per-atom array, and stash energy back into info
    atoms.new_array("force", forces)
    atoms.info["energy"] = energy

print("Arrays now:", list(train_atoms[0].arrays.keys()))
print("Info now:", list(train_atoms[0].info.keys()))

write("train_clean.xyz", train_atoms, format="extxyz")
write("test_clean.xyz", test_atoms, format="extxyz")

Arrays now: ['numbers', 'positions', 'force']
Info now: ['energy']


In [27]:
!sed -n '1,2p' train_clean.xyz

160
Lattice="9.7195997238 0.0 0.0 0.0 11.2231998444 0.0 0.0 0.0 50.0" Properties=species:S:1:pos:R:3:force:R:3 energy=-876.80328324 pbc="T T T"


In [28]:
import dpdata

ms_train = dpdata.MultiSystems.from_file(
    file_name="train_clean.xyz",
    fmt="mace/xyz",
    type_map=["H", "O", "Pt"]
)
print(ms_train)

MultiSystems (1 systems containing 80 frames)


In [29]:
ms_test = dpdata.MultiSystems.from_file(
    file_name="test_clean.xyz",
    fmt="mace/xyz",
    type_map=["H", "O", "Pt"]
)
print(ms_test)

MultiSystems (1 systems containing 21 frames)


In [31]:
import os

os.makedirs("processed_dpdata", exist_ok=True)

ms_train.to_deepmd_hdf5("processed_dpdata/train.h5")
ms_test.to_deepmd_hdf5("processed_dpdata/test.h5")

MultiSystems (1 systems containing 21 frames)

In [32]:
!find processed_dpdata -type f | sort

processed_dpdata/test.h5
processed_dpdata/train.h5


In [33]:
import h5py

def show(name, obj):
    if isinstance(obj, h5py.Dataset):
        print(f"{name}: shape={obj.shape}, dtype={obj.dtype}")

with h5py.File("processed_dpdata/train.h5", "r") as f:
    f.visititems(show)

H64O32Pt64/set.000/box.npy: shape=(80, 9), dtype=float64
H64O32Pt64/set.000/coord.npy: shape=(80, 480), dtype=float64
H64O32Pt64/set.000/energy.npy: shape=(80, 1), dtype=float64
H64O32Pt64/set.000/force.npy: shape=(80, 480), dtype=float64
H64O32Pt64/type.raw: shape=(160,), dtype=int64
H64O32Pt64/type_map.raw: shape=(3,), dtype=|S2


**Top-level group name:** H64O32Pt64 — this is dpdata's naming convention: it labels each distinct chemical formula as its own group. Since all 80 frames share the same composition (64 Pt, 32 O, 64 H — matching what we identified earlier as the Pt slab + water system), everything lives under one group. If your dataset had frames with varying atom counts/compositions, you'd see multiple sibling groups here.

**Inside** set.000/ (DeepMD's convention for a "set" — a batch/shard of frames with identical atom count and ordering):

- box.npy: shape (80, 9) — the 3×3 lattice/cell matrix flattened to 9 values, one row per frame. Matches your 80 training configs.
- coord.npy: shape (80, 480) — atomic positions, flattened. 480 = 160 atoms × 3 (x,y,z). One row per frame.
- energy.npy: shape (80, 1) — total energy per frame, matches the energy= values we saw in the raw xyz.
- force.npy: shape (80, 480) — forces, same flattening scheme as coords: 160 atoms × 3.

**Outside** set.000/, at the H64O32Pt64/ level:

- type.raw: shape (160,) — one integer per atom (not per frame), giving each atom's type index (0, 1, or 2, mapping into type_map.raw). This is fixed across all frames in this group since atom ordering matches.
- type_map.raw: shape (3,), dtype |S2 (byte-strings, max 2 chars) — the actual element symbols, e.g. [b'H', b'O', b'Pt'], defining what type index 0/1/2 means.

- **field_scalar** — the PE_MACE field-conditioning value, not representable in DeepMD's schema
- **stress** — DeepMD/dpdata can store stress/virial too, but we didn't request it and the header we wrote didn't include it (worth adding back if you need it — dpdata supports a virials.npy array if --stress data is present in the source)
- **dipole, magmom** — similarly dropped

In [34]:
import numpy as np

with h5py.File("processed_dpdata/train.h5", "r") as f:
    energies = f["H64O32Pt64/set.000/energy.npy"][:]
    type_map = f["H64O32Pt64/type_map.raw"][:]
    types = f["H64O32Pt64/type.raw"][:]

print("First frame energy:", energies[0])
print("Type map:", [t.decode() for t in type_map])
print("First 5 atom types:", types[:5])

First frame energy: [-876.80328324]
Type map: ['H', 'O', 'Pt']
First 5 atom types: [2 2 2 2 2]


In [35]:
with h5py.File("processed_dpdata/test.h5", "r") as f:
    def show(name, obj):
        if isinstance(obj, h5py.Dataset):
            print(f"{name}: shape={obj.shape}, dtype={obj.dtype}")
    f.visititems(show)

H64O32Pt64/set.000/box.npy: shape=(21, 9), dtype=float64
H64O32Pt64/set.000/coord.npy: shape=(21, 480), dtype=float64
H64O32Pt64/set.000/energy.npy: shape=(21, 1), dtype=float64
H64O32Pt64/set.000/force.npy: shape=(21, 480), dtype=float64
H64O32Pt64/type.raw: shape=(160,), dtype=int64
H64O32Pt64/type_map.raw: shape=(3,), dtype=|S2


MACE's mace_run_train HDF5 mode expects data preprocessed via mace_preprocess_data, which writes its own internal schema (grouped per-config with keys like positions, forces, atomic_numbers, etc. — different from DeepMD's set.000/coord.npy flattened-array-per-set layout we just built). MACE's HDF5 loader (mace/data/hdf5_dataset.py) does not know how to read DeepMD-formatted HDF5 — the internal key names and array shapes are fundamentally different formats that happen to share the .h5 container.

In [41]:
!find / -iname "PE_MACE_*" 2>/dev/null
!find / -iname "statistics.json" -newer train.xyz 2>/dev/null

/content/PE_MACE_statistics.json
/content/PE_MACE_val
/content/PE_MACE_train
/content/PE_MACE_test


In [42]:
!find /content/PE_MACE_train /content/PE_MACE_val /content/PE_MACE_test -type f | sort

/content/PE_MACE_test/Default__0.h5
/content/PE_MACE_train/train_0.h5
/content/PE_MACE_val/val_0.h5


In [43]:
!cat /content/PE_MACE_statistics.json

{"atomic_energies": "{1: np.float64(-6.092005843853211), 8: np.float64(-3.046002921926604), 78: np.float64(-6.092005843853206)}", "avg_num_neighbors": 41.20711898803711, "mean": -1.1067902050854173e-05, "std": 0.6367334723472595, "atomic_numbers": "[1, 8, 78]", "r_max": 5.0}

In [44]:
import h5py

def show(name, obj):
    if isinstance(obj, h5py.Dataset):
        print(f"{name}: shape={obj.shape}, dtype={obj.dtype}")
    else:
        print(f"{name}/")

# adjust filename once step 1 shows the real shard name
with h5py.File("/content/PE_MACE_train/train_0.h5", "r") as f:
    f.visititems(show)

config_batch_0/
config_batch_0/config_0/
config_batch_0/config_0/atomic_numbers: shape=(160,), dtype=int64
config_batch_0/config_0/cell: shape=(3, 3), dtype=float64
config_batch_0/config_0/config_type: shape=(), dtype=object
config_batch_0/config_0/pbc: shape=(3,), dtype=bool
config_batch_0/config_0/positions: shape=(160, 3), dtype=float64
config_batch_0/config_0/properties/
config_batch_0/config_0/properties/charges: shape=(), dtype=object
config_batch_0/config_0/properties/dipole: shape=(), dtype=object
config_batch_0/config_0/properties/energy: shape=(), dtype=float64
config_batch_0/config_0/properties/forces: shape=(160, 3), dtype=float64
config_batch_0/config_0/properties/head: shape=(), dtype=object
config_batch_0/config_0/properties/polarizability: shape=(), dtype=object
config_batch_0/config_0/properties/stress: shape=(), dtype=object
config_batch_0/config_0/properties/virials: shape=(), dtype=object
config_batch_0/config_0/property_weights/
config_batch_0/config_0/property_wei

In [45]:
with h5py.File("/content/PE_MACE_train/train_0.h5", "r") as f:
    def print_tree(name, indent=0):
        item = f[name]
        if isinstance(item, h5py.Group):
            print("  " * indent + f"{name.split('/')[-1]}/")
            for k in item.keys():
                print_tree(f"{name}/{k}" if name else k, indent + 1)
        else:
            val = item[()]
            preview = val if (hasattr(val, "size") and val.size < 10) else f"shape={item.shape}"
            print("  " * indent + f"{name.split('/')[-1]}: {preview}")
    for top_key in f.keys():
        print_tree(top_key)

config_batch_0/
  config_0/
    atomic_numbers: shape=(160,)
    cell: [[ 9.71959972  0.          0.        ]
 [ 0.         11.22319984  0.        ]
 [ 0.          0.         50.        ]]
    config_type: shape=()
    pbc: [ True  True  True]
    positions: shape=(160, 3)
    properties/
      charges: shape=()
      dipole: shape=()
      energy: -877.67206909
      forces: shape=(160, 3)
      head: shape=()
      polarizability: shape=()
      stress: shape=()
      virials: shape=()
    property_weights/
      charges: 0.0
      dipole: 0.0
      energy: 1.0
      forces: 1.0
      head: 1.0
      polarizability: 0.0
      stress: 0.0
      virials: 0.0
    weight: 1.0
  config_1/
    atomic_numbers: shape=(160,)
    cell: [[ 9.71959972  0.          0.        ]
 [ 0.         11.22319984  0.        ]
 [ 0.          0.         50.        ]]
    config_type: shape=()
    pbc: [ True  True  True]
    positions: shape=(160, 3)
    properties/
      charges: shape=()
      dipole: shape

In [47]:
import json
import re

with open("/content/PE_MACE_statistics.json", "r") as f:
    content = f.read()

# strip 'np.float64(...)' wrappers, keeping just the number inside
fixed_content = re.sub(r"np\.float64\(([^)]+)\)", r"\1", content)

with open("/content/PE_MACE_statistics_fixed.json", "w") as f:
    f.write(fixed_content)

print(fixed_content)

{"atomic_energies": "{1: -6.092005843853211, 8: -3.046002921926604, 78: -6.092005843853206}", "avg_num_neighbors": 41.20711898803711, "mean": -1.1067902050854173e-05, "std": 0.6367334723472595, "atomic_numbers": "[1, 8, 78]", "r_max": 5.0}


Run without Weights and Biases record, and this run did not evaluate the test data.

In [48]:
!mace_run_train \
    --name="MACE_Research_Intro_h5" \
    --train_file="/content/PE_MACE_train" \
    --valid_file="/content/PE_MACE_val" \
    --test_file="/content/PE_MACE_test" \
    --statistics_file="/content/PE_MACE_statistics_fixed.json" \
    --model="MACE" \
    --hidden_irreps="128x0e + 128x1o" \
    --r_max=5.0 \
    --batch_size=4 \
    --max_num_epochs=5 \
    --device=cuda

/usr/local/lib/python3.12/dist-packages/e3nn/o3/_wigner.py:10: UserWarning: Environment variable TORCH_FORCE_NO_WEIGHTS_ONLY_LOAD detected, since the`weights_only` argument was not explicitly passed to `torch.load`, forcing weights_only=False.
  _Jd, _W3j_flat, _W3j_indices = torch.load(os.path.join(os.path.dirname(__file__), 'constants.pt'))
/usr/local/lib/python3.12/dist-packages/torch/jit/_script.py:1488: DeprecationWarning: `torch.jit.script` is deprecated. Please switch to `torch.compile` or `torch.export`.
  warnings.warn(
2026-08-04 21:44:15.226 INFO: ===========VERIFYING SETTINGS===========
2026-08-04 21:44:15.226 INFO: MACE version: 0.3.17
2026-08-04 21:44:15.271 INFO: CUDA version: 12.8, CUDA device: 0
2026-08-04 21:44:15.308 INFO: ===========LOADING INPUT DATA===========
2026-08-04 21:44:15.309 INFO: Using heads: ['Default']
2026-08-04 21:44:15.309 INFO: Using the key specifications to parse data:
2026-08-04 21:44:15.309 INFO: Default: KeySpecification(info_keys={'energy': '

In [49]:
!mace_run_train \
    --name="MACE_Research_Intro_h5" \
    --train_file="/content/PE_MACE_train" \
    --valid_file="/content/PE_MACE_val" \
    --test_dir="/content/PE_MACE_test" \
    --multi_processed_test=True \
    --statistics_file="/content/PE_MACE_statistics_fixed.json" \
    --model="MACE" \
    --hidden_irreps="128x0e + 128x1o" \
    --r_max=5.0 \
    --batch_size=4 \
    --max_num_epochs=5 \
    --device=cuda \
    --wandb \
    --wandb_project="mace-research-project" \
    --wandb_name="MACE_Research_Intro_h5_run"

/usr/local/lib/python3.12/dist-packages/e3nn/o3/_wigner.py:10: UserWarning: Environment variable TORCH_FORCE_NO_WEIGHTS_ONLY_LOAD detected, since the`weights_only` argument was not explicitly passed to `torch.load`, forcing weights_only=False.
  _Jd, _W3j_flat, _W3j_indices = torch.load(os.path.join(os.path.dirname(__file__), 'constants.pt'))
/usr/local/lib/python3.12/dist-packages/torch/jit/_script.py:1488: DeprecationWarning: `torch.jit.script` is deprecated. Please switch to `torch.compile` or `torch.export`.
  warnings.warn(
2026-08-04 22:01:00.454 INFO: ===========VERIFYING SETTINGS===========
2026-08-04 22:01:00.454 INFO: MACE version: 0.3.17
2026-08-04 22:01:00.503 INFO: CUDA version: 12.8, CUDA device: 0
2026-08-04 22:01:00.544 INFO: ===========LOADING INPUT DATA===========
2026-08-04 22:01:00.544 INFO: Using heads: ['Default']
2026-08-04 22:01:00.544 INFO: Using the key specifications to parse data:
2026-08-04 22:01:00.544 INFO: Default: KeySpecification(info_keys={'energy': '

In [53]:
!sed -n '1065,1075p' /usr/local/lib/python3.12/dist-packages/mace/cli/run_train.py

                for test_file in test_files:
                    name = os.path.splitext(os.path.basename(test_file))[0]
                    test_sets[name] = data.HDF5Dataset(
                        test_file, r_max=args.r_max, z_table=z_table, heads=heads, head=head_config.head_name
                    )
            else:
                test_folders = glob.glob(head_config.test_dir + "/*")
                for folder in test_folders:
                    name = os.path.splitext(os.path.basename(test_file))[0]
                    test_sets[name] = data.dataset_from_sharded_hdf5(
                        folder, r_max=args.r_max, z_table=z_table, heads=heads, head=head_config.head_name


In [54]:
path = "/usr/local/lib/python3.12/dist-packages/mace/cli/run_train.py"

with open(path, "r") as f:
    content = f.read()

old_block = '''for folder in test_folders:
                    name = os.path.splitext(os.path.basename(test_file))[0]
                    test_sets[name] = data.dataset_from_sharded_hdf5('''

new_block = '''for folder in test_folders:
                    name = os.path.splitext(os.path.basename(folder))[0]
                    test_sets[name] = data.dataset_from_sharded_hdf5('''

assert old_block in content, "Pattern not found — check exact whitespace/indentation"
content = content.replace(old_block, new_block)

with open(path, "w") as f:
    f.write(content)

print("Second bug patched")

Second bug patched


In [55]:
!sed -n '1069,1080p' /usr/local/lib/python3.12/dist-packages/mace/cli/run_train.py

                    )
            else:
                test_folders = glob.glob(head_config.test_dir + "/*")
                for folder in test_folders:
                    name = os.path.splitext(os.path.basename(folder))[0]
                    test_sets[name] = data.dataset_from_sharded_hdf5(
                        folder, r_max=args.r_max, z_table=z_table, heads=heads, head=head_config.head_name
                    )
        for test_name, test_set in test_sets.items():
            test_sampler = None
            if args.distributed:
                test_sampler = torch.utils.data.distributed.DistributedSampler(


In [57]:
!mace_run_train \
    --name="MACE_Research_Intro_h5" \
    --train_file="/content/PE_MACE_train" \
    --valid_file="/content/PE_MACE_val" \
    --test_dir="/content/PE_MACE_test" \
    --multi_processed_test=False \
    --statistics_file="/content/PE_MACE_statistics_fixed.json" \
    --model="MACE" \
    --hidden_irreps="128x0e + 128x1o" \
    --r_max=5.0 \
    --batch_size=4 \
    --max_num_epochs=5 \
    --device=cuda \
    --wandb \
    --wandb_project="mace-research-project" \
    --wandb_name="MACE_Research_Intro_h5_run2"

/usr/local/lib/python3.12/dist-packages/e3nn/o3/_wigner.py:10: UserWarning: Environment variable TORCH_FORCE_NO_WEIGHTS_ONLY_LOAD detected, since the`weights_only` argument was not explicitly passed to `torch.load`, forcing weights_only=False.
  _Jd, _W3j_flat, _W3j_indices = torch.load(os.path.join(os.path.dirname(__file__), 'constants.pt'))
/usr/local/lib/python3.12/dist-packages/torch/jit/_script.py:1488: DeprecationWarning: `torch.jit.script` is deprecated. Please switch to `torch.compile` or `torch.export`.
  warnings.warn(
2026-08-04 22:17:13.557 INFO: ===========VERIFYING SETTINGS===========
2026-08-04 22:17:13.558 INFO: MACE version: 0.3.17
2026-08-04 22:17:13.608 INFO: CUDA version: 12.8, CUDA device: 0
2026-08-04 22:17:13.648 INFO: ===========LOADING INPUT DATA===========
2026-08-04 22:17:13.648 INFO: Using heads: ['Default']
2026-08-04 22:17:13.648 INFO: Using the key specifications to parse data:
2026-08-04 22:17:13.648 INFO: Default: KeySpecification(info_keys={'energy': '

# MACE Training & Evaluation — Working Reference

Distilled from a working session on MACE 0.3.17. Covers both the direct-XYZ
path and the HDF5-preprocessing path, plus every gotcha that actually broke
something.

---

## 0. Data format requirements (apply to BOTH paths)

Your extended-XYZ file's `Properties=` header must contain **exactly**:
```
Properties=species:S:1:pos:R:3:forces:R:3
```
- Use `forces` (plural) — this is what MACE's own reader expects.
- Any extra per-atom field not in {`species`, `pos`, `forces`} (e.g. a custom
  `field_scalar` column) is fine for MACE's own tools, but will **hard-crash**
  dpdata or QUIP/GAP-style parsers. Strip such columns before using anything
  outside the MACE ecosystem.
- Global key=value line must include `energy=...`. Don't rely on `energy`
  as the key name long-term — ASE/MACE warn that `energy`/`forces` are
  "unsafe" going forward; prefer `REF_energy`/`REF_forces` and pass
  `--energy_key=REF_energy --forces_key=REF_forces` if starting fresh.
- Download raw files with `raw.githubusercontent.com`, never a GitHub
  `blob/` URL — the latter downloads an HTML page, not the data.

---

## 1. Direct XYZ training (simplest path, good for quick tests)

```bash
mace_run_train \
    --name="RUN_NAME" \
    --train_file="train.xyz" \
    --test_file="test.xyz" \
    --valid_fraction=0.1 \
    --energy_key="energy" \
    --forces_key="forces" \
    --E0s="average" \
    --model="MACE" \
    --hidden_irreps="128x0e + 128x1o" \
    --r_max=5.0 \
    --batch_size=4 \
    --max_num_epochs=500 \
    --device=cuda \
    --wandb --wandb_project="PROJECT_NAME" --wandb_name="RUN_NAME"
```

**Must-dos:**
- Always include `--test_file` if you want real held-out test metrics —
  without it, MACE only reports train/valid (validation is just a random
  10% split of train, not a true test set).
- `--max_num_epochs=5` is a smoke test, not a real run. Use hundreds+ and
  watch wandb/loss curves for a plateau before trusting results.
- No trailing spaces after line-continuation `\` — a stray space causes
  `unrecognized arguments: \`.
- In Jupyter/Colab, prefix shell commands with `!`.

**Evaluating an already-trained model on new data (no retraining):**
```bash
mace_eval_configs \
    --configs="test.xyz" \
    --model="checkpoints/RUN_NAME_run-123.model" \
    --output="test_eval.xyz" \
    --device=cuda \
    --batch_size=4
```
- No `--energy_key`/`--forces_key` here — not accepted by this tool.
- If CUDA OOM: lower `--batch_size` (try 4, 2, 1) before anything else.

---

## 2. HDF5 preprocessing + training (for larger/repeated-use datasets)

### Step 1 — Preprocess (module has no console-script entry point in 0.3.17):
```bash
python -m mace.cli.preprocess_data \
    --train_file="train.xyz" \
    --valid_fraction=0.1 \
    --test_file="test.xyz" \
    --h5_prefix="PREFIX_" \
    --r_max=5.0 \
    --energy_key="energy" \
    --forces_key="forces" \
    --atomic_numbers="[Z1, Z2, Z3]" \
    --compute_statistics \
    --E0s="average" \
    --num_process=1 \
    --seed=123
```
**Gotchas:**
- `--work_dir` is effectively ignored in this version — output lands in the
  current directory as `PREFIX_train/`, `PREFIX_val/`, `PREFIX_test/`,
  `PREFIX_statistics.json`. Locate with `find / -iname "PREFIX_*"` if unsure.
- `--atomic_numbers` must be given explicitly if E0s aren't in the file.

### Step 2 — Fix the statistics.json numpy-serialization bug:
The tool writes `np.float64(...)` wrapped inside a JSON *string*, which
`ast.literal_eval` (used by `mace_run_train`) cannot parse. Patch before use:
```python
import re
with open("PREFIX_statistics.json") as f:
    content = f.read()
content = re.sub(r"np\.float64\(([^)]+)\)", r"\1", content)
with open("PREFIX_statistics_fixed.json", "w") as f:
    f.write(content)
```

### Step 3 — Patch two known bugs in `mace/cli/run_train.py` (0.3.17 only, check if fixed in newer versions):
```python
path = "/usr/local/lib/python3.12/dist-packages/mace/cli/run_train.py"
with open(path) as f: c = f.read()
c = c.replace(
    'test_folders = glob(head_config.test_dir + "/*")',
    'test_folders = glob.glob(head_config.test_dir + "/*")')
c = c.replace(
    'for folder in test_folders:\n                    name = os.path.splitext(os.path.basename(test_file))[0]',
    'for folder in test_folders:\n                    name = os.path.splitext(os.path.basename(folder))[0]')
with open(path, "w") as f: f.write(c)
```
(Bug 1: `glob` imported as module but called as function. Bug 2: stale loop
variable `test_file` used instead of `folder`.)

### Step 4 — Train from HDF5:
```bash
mace_run_train \
    --name="RUN_NAME" \
    --train_file="/path/PREFIX_train" \
    --valid_file="/path/PREFIX_val" \
    --test_file="/path/PREFIX_test/Default__0.h5" \
    --statistics_file="/path/PREFIX_statistics_fixed.json" \
    --model="MACE" \
    --hidden_irreps="128x0e + 128x1o" \
    --r_max=5.0 \
    --batch_size=4 \
    --max_num_epochs=500 \
    --device=cuda \
    --wandb --wandb_project="PROJECT_NAME" --wandb_name="RUN_NAME"
```
**Must-dos:**
- `--train_file` / `--valid_file` point at **folders** of shards.
- `--test_file` points at the actual **single `.h5` file** (e.g.
  `Default__0.h5`) directly — NOT `--test_dir` +
  `--multi_processed_test=True`, which is only for a directory containing
  *subfolders* of shards (a different, sharded-multi-file layout you likely
  won't have unless you used `--num_process` > 1 during preprocessing to
  generate multiple shards, or an actual multi-shard dataset).
- With `--statistics_file` given, drop `--E0s`/`--energy_key`/`--forces_key`
  — they're already baked into the HDF5 labels.
- Test shard filenames are prefixed by head name + double underscore, e.g.
  `Default__0.h5` — check via `find PREFIX_test -type f` rather than
  assuming a name.

---

## 3. Do NOT use dpdata for MACE HDF5 (dead end, confirmed this session)

dpdata produces DeepMD-kit-schema HDF5 (`set.000/coord.npy` style), which
`mace_run_train` cannot read — the internal key names and layout are
fundamentally different from MACE's own `config_batch_0/config_N/...`
schema. Only use dpdata if you're actually training DeepMD-kit, not MACE.
If you need MACE HDF5, always go through `mace.cli.preprocess_data`.

---

## 4. General sanity-check habits worth keeping

- After any format conversion, spot-check one energy value against the
  original raw XYZ (`grep energy= file.xyz | head -1`) to confirm no
  silent data loss/reordering.
- After training, always check the RESULTS section actually printed an
  `Evaluating test_Default ...` line — its absence means the test set was
  silently skipped, not that it passed with zero error.
- 5 epochs / ~80 configs is a "does the pipeline run" smoke test only.
  Expect RMSE_E in the hundreds of meV/atom at that scale — not a real
  quality signal. For real DFT data, plan for hundreds–thousands of epochs
  and a dataset several times larger before trusting reported errors.